server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -514,97 +514,121 @@ TimeValue prepareTargetForTranslog(final boolean fileBasedRecovery, final int to
-        ActionListener.completeWith(listener, () -> sendSnapshotBlockingly(
-            startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, snapshot, maxSeenAutoIdTimestamp, maxSeqNoOfUpdatesOrDeletes));
-    }
-
-    private SendSnapshotResult sendSnapshotBlockingly(long startingSeqNo, long requiredSeqNoRangeStart, long endingSeqNo,
-                                                      Translog.Snapshot snapshot, long maxSeenAutoIdTimestamp,
-                                                      long maxSeqNoOfUpdatesOrDeletes) throws IOException {
         assert requiredSeqNoRangeStart <= endingSeqNo + 1:
             "requiredSeqNoRangeStart " + requiredSeqNoRangeStart + " is larger than endingSeqNo " + endingSeqNo;
         assert startingSeqNo <= requiredSeqNoRangeStart :
             "startingSeqNo " + startingSeqNo + " is larger than requiredSeqNoRangeStart " + requiredSeqNoRangeStart;
         if (shard.state() == IndexShardState.CLOSED) {
             throw new IndexShardClosedException(request.shardId());
         }
-
-        final StopWatch stopWatch = new StopWatch().start();
-
         logger.trace("recovery [phase2]: sending transaction log operations (seq# from [" +  startingSeqNo  + "], " +
             "required [" + requiredSeqNoRangeStart + ":" + endingSeqNo + "]");
 
-        int ops = 0;
-        long size = 0;
-        int skippedOps = 0;
-        int totalSentOps = 0;
-        final AtomicLong targetLocalCheckpoint = new AtomicLong(SequenceNumbers.UNASSIGNED_SEQ_NO);
-        final List<Translog.Operation> operations = new ArrayList<>();
+        final AtomicInteger skippedOps = new AtomicInteger();
+        final AtomicInteger totalSentOps = new AtomicInteger();
         final LocalCheckpointTracker requiredOpsTracker = new LocalCheckpointTracker(endingSeqNo, requiredSeqNoRangeStart - 1);
 
-        final int expectedTotalOps = snapshot.totalOperations();
-        if (expectedTotalOps == 0) {
-            logger.trace("no translog operations to send");
-        }
-
-        final CancellableThreads.IOInterruptible sendBatch = () -> {
-            // TODO: Make this non-blocking
-            final PlainActionFuture<Long> future = new PlainActionFuture<>();
-            recoveryTarget.indexTranslogOperations(
-                operations, expectedTotalOps, maxSeenAutoIdTimestamp, maxSeqNoOfUpdatesOrDeletes, future);
-            targetLocalCheckpoint.set(future.actionGet());
+        // Wrap translog snapshot to make it synchronized as it is accessed by different threads through sendBatch.
+        // Even though those calls are not concurrent, snapshot.next() uses non-synchronized state and is not multi-thread-compatible.
+        final Translog.Snapshot wrappedSnapshot = synchronizedSnapshot(snapshot);
+        final CheckedSupplier<List<Translog.Operation>, IOException> readNextBatch = () -> {
+            final List<Translog.Operation> operations = new ArrayList<>();
```

Comment:
can we maybe estimate the size of the array list? it would be nice to somehow can carry over the size of the last batch?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/37503/commits/1a7e4022571226086f0d250347a4ec4371e6d9e3.